src/units.rs

```diff
@@ -0,0 +1,385 @@
+use crate::{
+    member::NotificationOut, Data, Hash, KeyBox, NodeCount, NodeIndex, NodeMap, Round, SessionId,
+};
+use codec::{Decode, Encode};
+use log::{debug, error};
+use std::{collections::HashMap, hash::Hash as StdHash};
+
+// TODO: need to make sure we never accept units of round > MAX_ROUND
+pub(crate) const MAX_ROUND: usize = 5000;
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub(crate) struct FullUnit<H: Hash, D: Data> {
+    pub(crate) inner: PreUnit<H>,
+    pub(crate) data: D,
+    pub(crate) session_id: SessionId,
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub(crate) struct SignedUnit<H: Hash, D: Data, Signature: Clone + Encode + Decode> {
+    pub(crate) unit: FullUnit<H, D>,
+    pub(crate) signature: Signature,
+    pub(crate) id: NodeIndex,
+}
+
+impl<H: Hash, D: Data, Signature: Clone + Encode + Decode> SignedUnit<H, D, Signature> {
+    /// Verifies the unit's signature. The signature is verified on creation, so this should always
+    /// return true, but the method can be used to check integrity.
+    pub(crate) fn verify_signature<KB: KeyBox<Signature>>(&self, keybox: &KB) -> bool {
+        if !keybox.verify(&self.unit.encode(), &self.signature) {
+            debug!(target: "rush-member", "Bad signature in a unit from {:?}", self.unit.inner.creator());
+            return false;
+        }
+        true
+    }
```

Comment:
As mentioned in a comment above: I don't like this KeyBox API, because it enforces the signature to carry the index, which I think is inefficient.